### PR TITLE
API for database config 

### LINKED ISSUE #3157: API for adding/modifying user database credentials in Mathesar
This issue is for tracking the backend work needed for implementing the design in #3159

## Proposed solution
- Create API to add the user database credentials in the internal Django database
  - The following information will be sent from the client
    - Unique name for the database
    - Database name
    - Database username
    - Database password
    - Database host
    - Database port
  - The database credentials other than the unique name should be encrypted using the `SECRET_KEY` when stored in the database and decrypted back using the same `SECRET_KEY` when decrypting back
  - Mathesar schema (containing Mathesar types and schema) should be automatically installed into the user database when it is added to Mathesar
  - Can be added only by the superuser
 
- Create API to edit the user database credentials in the internal Django database
  - The following information can be edited
    - Database name
    - Database username
    - Database password
    - Database host
    - Database port
  - Can be edited only by the superuser
  
- Create API to remove the user database credentials
  - Optionally, accept boolean to remove Mathesar Schema when the user database credential is removed from Mathesar
  - If the boolean to remove Mathesar schema is `True`, we should remove Mathesar schema installed on the database first before removing the credentials from Mathesar
  - Can be removed only by the superuser

### LINKED ISSUE #3158: Update the database objects created using the env to be consistent with the database objects created using the API
Follow-up of #3157  Database objects added using the environment variable `MATHESAR_DATABASES` should also include `username`, `password`, `hostname` and `port` to be consistent with the database objects created using the database API. These databases objects added using the environment variable should also be read-only and should not be editable using the API.

## Additional context

- [This is the line](https://github.com/centerofci/mathesar/blob/8018a4404301e896b95ffe857acb55669456a434/mathesar/apps.py#L13) which adds the environment variables

### PR DESCRIPTION
**Pull request by Anish9901**
Fixes #3157
Fixes #3158

**Technical details**

API Signature:
- Create DB credentials:
```python
POST /api/db/v0/databases/
```
```sql
{
    "name": "mathesar", (unique)
    "db_username": "mathesar",
    "db_password": "mathesar",
    "db_host": "mathesar_dev_db",
    "db_port": 5432
}
```
- Alter DB credentials:
```python
PATCH /api/db/v0/databases/<id>
```
```sql
{
    "name": "mathesar", (unique)
    "db_username": "mathesar",
    "db_password": "mathesar",
    "db_host": "mathesar_dev_db",
    "db_port": 5432
}
```
- Delete DB credentials:
```python
DELETE /api/db/v0/databases/<id>
```
- Delete DB credentials & msar SQL & types schemas:
```python
DELETE /api/db/v0/databases/<id>/?del_msar_schemas=True
```
Changed `common_data` dict:
- If someone edits their database credentials outside of mathesar the django model for that connection becomes stale.
- To address this the `databases` key in the `common_dict` will now provide a dict something like this for the corresponding stale database object:
```sql
{
    'name': 'mt_tables',
    'editable': True,
    'error': 'Error connecting to the database'
}
``` 

## Checklist
 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `develop` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [ ] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by Anish9901** [comment]
There is a issue with the install script that I'd like to point out, whenever there is a change in the function signature of our SQL functions the install script executes without termination even though psycopg exceptions can be noticed in the logs.

I don't think it's a good idea to reinstall those schema without the consent of the users as there might be db objects dependent on those schemas.

File: mathesar/api/db/viewsets/databases.py
Comment (silentninja) [inline mathesar/api/db/viewsets/databases.py:41]: It would be better to perform this validation inside the serializer.

File: mathesar/api/serializers/databases.py
Comment (silentninja) [inline mathesar/api/serializers/databases.py:11]: Is there a reason to explicitly specify the field types here, Doesn't drf infer from the model? It would be good to add comments on the need for overriding the field type

**Review by silentninja — CHANGES_REQUESTED** [review]
This is a good start! However, there are a few issues

- The Database model does not contain the unique name field. We cannot use the database name as a reference as it can be changed too even though the data inside the database is the same. Please add it to the `Database` model.

- I have made specific comments to highlight other issues

**Comment by silentninja** [comment]
> whenever there is a change in the function signature of our SQL functions the install script executes without termination even though psycopg exceptions can be noticed in the logs.

I thought we decided not to have a breaking function signature between release. If you are noticing this in the `develop` branch I don't think it is a problem. If not please raise an issue for it

> I don't think it's a good idea to reinstall those schema without the consent of the users as there might be db objects dependent on those schemas.

For the sake of easier upgrades, we decided to not have any breaking changes. So I don't think it will affect any dependent objects

**Comment by Anish9901** [comment]
> The Database model does not contain the unique name field. We cannot use the database name as a reference as it can be changed too even though the data inside the database is the same. Please add it to the Database model.

`unique` is set to `True` in the Database model for the `name` field, am I missing something?

File: mathesar/exception_handlers.py
Comment (Anish9901) [reply mathesar/exception_handlers.py]: [This](https://github.com/centerofci/mathesar/blob/00204a7198cd21351c121e739fcd76014790942b/mathesar/api/exceptions/validation_exceptions/base_exceptions.py#L21) not being a list was causing `is_pretty()` to fail, fixed now!

**Comment by silentninja** [comment]
> unique is set to True in the Database model for the name field, am I missing something?

If the database object is added using a .env file, the only way to track the database object across database name change is by using a unique database key which is not tied to the property of the database. So we need a unique database key on top of database name

**Review by silentninja — APPROVED** [review]
Thanks @Anish9901, it looks good to me! I would like to wait until the frontend work is done to merge this PR as it will be easier to confirm if something is broken.

**Pull request merged**
